neuralpp/symbolic/sympy_expression.py

```diff
@@ -1,75 +1,121 @@
 from __future__ import annotations
-from typing import List, Any, Type, Callable, Tuple
+from typing import List, Any, Type, Callable, Tuple, Optional, Dict
 from abc import ABC
 
 import sympy
 import operator
 import builtins
-from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant
+import fractions
+from neuralpp.symbolic.expression import Expression, FunctionApplication, Variable, Constant, \
+    FunctionNotTypedError, NotTypedError, FunctionType
+from neuralpp.symbolic.basic_expression import BasicConstant, new_type, infer_python_callable_type
+from neuralpp.util.util import consistently_update_dict
+
 
 # In this file's doc, I try to avoid the term `sympy expression` because it could mean both sympy.Expr (or sympy.Basic)
 # and SymPyExpression. I usually use "sympy object" to refer to the former and "expression" to refer to the latter.
 
 
-def python_callable_to_sympy_function(python_callable: Callable) -> Type[sympy.Basic]:
-    match python_callable:
-        # boolean operation
-        case operator.__and__:
-            return sympy.And
-        case operator.__or__:
-            return sympy.Or
-        case operator.__not__:
-            return sympy.Not
-        case operator.__xor__:
-            return sympy.Xor
-        # comparison
-        case operator.le:
-            return sympy.Le
-        case operator.lt:
-            return sympy.Lt
-        case operator.ge:
-            return sympy.Ge
-        case operator.gt:
-            return sympy.Gt
-        case operator.eq:
-            return sympy.Eq
-        # arithmetic
-        case operator.add:
-            return sympy.Add
-        case operator.mul:
-            return sympy.Mul
-        case operator.pow:
-            return sympy.Pow
-        # min/max
-        case builtins.min:
-            return sympy.Min
-        case builtins.max:
-            return sympy.Max
+def infer_sympy_object_type(sympy_object: sympy.Basic, type_dict: Dict[sympy.Basic, Expression]) -> Expression:
+    match sympy_object:
+        case sympy.Integer():
+            return BasicConstant(int)
+        case sympy.Float():
+            return BasicConstant(float)
+        case sympy.Rational():
+            return BasicConstant(fractions.Fraction)
+        case sympy.logic.boolalg.BooleanAtom():
+            return BasicConstant(bool)
         case _:
-            raise ValueError(f"Python callable {python_callable} is not recognized.")
+            # It is obvious that we can look up type_dict for symbols like `symbol("x")`.
+            # We can also look up for function, such as `a+b`, where type_dict records the function type (instead of
```

Comment:
I didn't quite understand this comment. You write "for function, such as `a+b`", but that is a function application, not a function, so I got confused. It would be helpful to write the comment for `infer_sympy_object_type` telling us what `type_dict` is supposed to contain exactly.
When you say "where type_dict records the function type (instead of the return type)", that sounds like you are defining what `type_dict` is supposed to do/be, in which case that would be better as part of the docstring of the method.
I also got confused by what you meant by "root-level function". Is that the same as "the function used in the function application"? If so, using that would probably be easier to understand.